opamp/observiq/observiq_client.go

```diff
@@ -334,230 +325,274 @@ func (c *Client) onPackagesAvailableHandler(packagesAvailable *protobufs.Package
 	if c.safeGetUpdatingPackage() {
 		c.logger.Warn(
 			"Not starting new package update as already installing new packages",
-			zap.String("AllPackagesHash", hex.EncodeToString(packagesAvailable.GetAllPackagesHash())))
-		curPackageStatuses.ErrorMessage = "Already installing new packages"
-		if err := c.opampClient.SetPackageStatuses(curPackageStatuses); err != nil {
-			c.logger.Error("OpAMP client failed to set package statuses", zap.Error(err))
+			zap.String("AllPackagesHash", hex.EncodeToString(availablePkgs.GetAllPackagesHash())))
+
+		curPkgStatuses.ErrorMessage = "Already installing new packages"
+		// Dont' actually set the on file package statuses because we want to ignore this
+		if err := c.opampClient.SetPackageStatuses(curPkgStatuses); err != nil {
+			c.logger.Error("OpAMP client failed to set already installing package statuses", zap.Error(err))
 		}
-		return fmt.Errorf("failed because already installing packages")
+		return errors.New("failed because already installing packages")
 	}
 
-	// Retrieve last known status (this should return with minimal info even on first time)
-	lastPackageStatuses, err := c.packagesStateProvider.LastReportedStatuses()
-
+	// Retrieve last known status (this should return with minimal info even on first time).
 	// If there is a problem retrieving the last saved PackageStatuses, we will log the error
 	// but continue on as the only thing missing will be the agent package hash.
+	pkgStatuses, err := c.getVerifiedPackageStatuses()
 	if err != nil {
-		c.logger.Warn("Failed to retrieve last reported package statuses", zap.Error(err))
+		c.logger.Warn("Problem with package statuses on starting install", zap.Error(err))
 	}
 
-	var lastPkgStatusMap map[string]*protobufs.PackageStatus
-	if lastPackageStatuses != nil {
-		lastPkgStatusMap = lastPackageStatuses.GetPackages()
+	lastPkgStatusMap := make(map[string]*protobufs.PackageStatus)
+	if pkgStatuses != nil && pkgStatuses.GetPackages() != nil {
+		lastPkgStatusMap = pkgStatuses.GetPackages()
 	}
 
-	curPackages, curPackageFiles := c.createPackageMaps(packagesAvailable.GetPackages(), lastPkgStatusMap)
-	curPackageStatuses.Packages = curPackages
+	// Loop through all of the available packages sent from the server and create initial PackageStatuses
+	for pkgName, availablePkg := range availablePkgs.Packages {
+		lastPkgStatus := lastPkgStatusMap[pkgName]
+		curPkgStatuses.Packages[pkgName] = c.buildInitialPackageStatus(pkgName, availablePkg, lastPkgStatus)
+	}
 
 	// This is an error because we need this file for communication during the update
-	if err = c.packagesStateProvider.SetLastReportedStatuses(curPackageStatuses); err != nil {
+	if err = c.packagesStateProvider.SetLastReportedStatuses(curPkgStatuses); err != nil {
 		return fmt.Errorf("failed to save last reported package statuses: %w", err)
 	}
```

Comment:
This is kinda unrelated to the refactor, BUT I noticed that we fail here without sending package statuses back to bindplane; Should we try to send a failure message back in this case, or is this intentional?

Reply:
I think yes and no? If we can't write to package statuses but we send status (success or failure) to bindplane, if we were to shut down and start up we'd have a bad last status. Maybe it's not the worst thing in the world and would possibly self correct? I don't know if there is a "right" answer here.